Python/Product/Uwp/Project/PythonUwpProjectConfig.cs

```diff
@@ -920,31 +979,43 @@ class PythonUwpProjectConfig :
             return value.Equals("true", StringComparison.OrdinalIgnoreCase);
         }
 
-        private int Deploy() {
-            try {
-                foreach (IVsDeployStatusCallback callback in deployCallbackCollection) {
-                    callback.OnEndDeploy(1);
+        private bool NotifyBeginDeploy() {
+            foreach (IVsDeployStatusCallback callback in GetSinkCollection()) {
+                if (callback == null) {
+                    continue;
                 }
-            } finally {
-                lock (syncObject) {
-                    appContainerBootstrapperOperation = null;
-                    deployOp = null;
+
+                int fContinue = 1;
+
+                if (ErrorHandler.Failed(callback.OnStartDeploy(ref fContinue)) || fContinue == 0) {
+                    return false;
                 }
             }
 
-            this.outputWindow = null;
-
-            return VSConstants.S_OK;
+            return true;
         }
 
         private void NotifyEndDeploy(int success) {
-            foreach (IVsDeployStatusCallback callback in deployCallbackCollection) {
-                callback.OnEndDeploy(success);
+            try {
+                foreach (IVsDeployStatusCallback callback in GetSinkCollection()) {
+                    callback.OnEndDeploy(success);
+                }
+            } finally {
+                lock (syncObject) {
+                    this.appContainerBootstrapperOperation = null;
+                    this.deployOp = null;
+                }
             }
 
             outputWindow = null;
         }
 
+        private IEnumerable<IVsDeployStatusCallback> GetSinkCollection() {
+            lock (syncObject) {
+                return this.deployCallbackCollection.Cast<IVsDeployStatusCallback>().ToList();
```

Comment:
Using `OfType<T>` instead of `Cast<T>` will remove nulls from the resulting sequence.